service/controller/resource/instance/status.go

```diff
@@ -1,86 +1,67 @@
 package instance
 
 import (
-	"strings"
+	"context"
 
-	providerv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"sigs.k8s.io/cluster-api-provider-azure/exp/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/pkg/annotation"
 )
 
 const (
 	// Types
-	Stage                        = "Stage"
-	DeploymentTemplateChecksum   = "TemplateChecksum"
-	DeploymentParametersChecksum = "ParametersChecksum"
+	Stage = "Stage"
 
 	// States
-	ClusterUpgradeRequirementCheck = "ClusterUpgradeRequirementCheck"
-	CordonOldWorkers               = "CordonOldWorkers"
-	DeploymentUninitialized        = ""
-	DeploymentInitialized          = "DeploymentInitialized"
-	DeploymentCompleted            = "DeploymentCompleted"
-	DrainOldWorkerNodes            = "DrainOldWorkerNodes"
-	ProvisioningSuccessful         = "ProvisioningSuccessful"
-	ScaleUpWorkerVMSS              = "ScaleUpWorkerVMSS"
-	ScaleDownWorkerVMSS            = "ScaleDownWorkerVMSS"
-	TerminateOldWorkerInstances    = "TerminateOldWorkerInstances"
-	WaitForWorkersToBecomeReady    = "WaitForWorkersToBecomeReady"
+	CordonOldWorkers            = "CordonOldWorkers"
+	DeploymentUninitialized     = ""
+	DrainOldWorkerNodes         = "DrainOldWorkerNodes"
+	ScaleUpWorkerVMSS           = "ScaleUpWorkerVMSS"
+	ScaleDownWorkerVMSS         = "ScaleDownWorkerVMSS"
+	TerminateOldWorkerInstances = "TerminateOldWorkerInstances"
+	WaitForWorkersToBecomeReady = "WaitForWorkersToBecomeReady"
 )
 
-func computeForDeleteResourceStatus(customObject providerv1alpha1.AzureConfig, n string, t string, s string) providerv1alpha1.AzureConfig {
-	var allResources []providerv1alpha1.StatusClusterResource
-
-	for _, r := range customObject.Status.Cluster.Resources {
-		resourceStatus := providerv1alpha1.StatusClusterResource{
-			Conditions: nil,
-			Name:       n,
-		}
-
-		// At this point we ensure resource statuses of other resources are
-		// preserved as they are. When we want to remove a status of resource A, but
-		// find a status of resource B, we keep B, because we were not asked to
-		// remove it.
-		if !unversionedNameMatches(r.Name, n) {
-			allResources = append(allResources, r)
-			continue
-		}
+func (r *Resource) saveCurrentState(ctx context.Context, customObject v1alpha3.AzureMachinePool, state string) error {
```

Comment:
The state machine current state is saved in an annotation inside `AzureMachinePool`.